### ISSUE TITLE
webrtc.js:76 WebSocket connection to 'wss://127.0.0.1/' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED

### ISSUE THREAD

**Issue by yarekc** [akshat7101999/WebRTC-scalable-broadcast-p2p#3]
I tun it on win
npm start
Then I go to : 
https://127.0.0.1:8443/

https://prnt.sc/tykj1b

Can see my cam and got that error

webrtc.js:76 WebSocket connection to 'wss://127.0.0.1/' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED

Any idea how to fix that ?

**Comment by akshat7101999** [comment]
Hey yarekc ....... Well comment line number 76 and uncomment line 75 ....and replace my ip in line 75 with 127.0.0.1:8443 ... and i guess it will solve the issue . Right now it is unable to form connection with server. You can reach out again if the issue persists. Also url on your tab should be https://127.0.0.1:8443 (it shoould be https not http)

**Comment by yarekc** [comment]
thanks. Now... I really don't unerstand how it works (or maybe it does not work)
user1 connects : he sees himself
user2 connects : he sees himself 

in 90% of cases, NOTHING else happens. In 10% of cases, user1 sees user2 and user2 sees user1 
It is random and mostly of time nothing happens. Sometimes, conference starts.
I have no JS errors, no server errors.

When user3 joins: 
- user3 can see user3 and user2
- user2 can see user2 and user3
- user1 sees user1

It is however very random and I never success in having correct result. 

The raise hand has no effect at all, except printing "undefined" in the node console.

My question is: Is that working ? 
If yes, I  suggest you publish a demo online

**Comment by akshat7101999** [comment]
Do you have latency issues ?  Because if page takes more than 2 seconds to load then it breaks and can be modified in set-timeout function at line 81. Also i see you are getting random results. It occurs if you are a low bandwidth client. Are you getting an alert message of *You are running on low bandwidth* ??

Well algorithm should work like this - >

user1 sees user1(user 1 is the master)
user2 can see user2(own stream) and user1
user3 can see user3 and user1
user4 can see user4 and user2
user5 can see user5 and user3
user6 can see user6 and user4
user7 can see user7 and user5

basically a branched chain if formed.

See my working demo

https://detectronlive123.000webhostapp.com/1.mp4

https://detectronlive123.000webhostapp.com/2.mp4

The algorithm is as follows - 
https://detectronlive123.000webhostapp.com/algo.png

**Comment by yarekc** [comment]
I have a high speed fiber connection. 
Infact I think don't understand the purpose of this project. What is the purpose user7 sees user 7 and user 5 ? (And not other users ?) In real life?

I think it can be useful if any user can see any other user he decides.

Regards

**Comment by akshat7101999** [comment]
This was my research intern project at IIT Kanpur. You dont have to see any other user in the broadcast. What exactly happening here is that master's feed should be visible to all (since master is the one broadcasting his feed). Now **master is sending its stream to user 2. user4 is getting the feed of master from user 2 (relaying of streams). User 6 is getting master's stream from user 4 and so on.** This way there is no load on master and p2p streaming is taking place. You should read about the algorithm here - https://docs.google.com/presentation/d/1YLxouDWRVHrKxgG6sKfOq8Ge4pQXBH7tD2rU1WpY1Yo/edit?usp=sharing
This is quite scalable. We made this so that teacher can teach a class of 100 students. Students need not be able to see themselves but everyone will see teacher's stream.

If you want that everyone should see everyone then follow - https://github.com/akshat7101999/WebRTC-mesh-group-video-call
It is in no way scalable and with 5 6 users in video conferencing your bandwidth will be exhausted and so will be you cpu

**Comment by yarekc** [comment]
I understand the concept of mesh, peer2peer and scalable broadcast.
Here is what you write: 
user1 sees user1(user 1 is the master)
user2 can see user2(own stream) and user1
user3 can see user3 and user1
user4 can see user4 and user2
user5 can see user5 and user3
user6 can see user6 and user4
user7 can see user7 and user5

**Question is: WHY all these users do not see the user1 (the master user) ?**

They should ALL be able to use user1 otherwise what is the purpose of having a mesh like
user19 is only able to see user598595 and user5455 only to see user511255 : for me that has no sense.

I guess the Goal of the project is that ALL users are able to see the SAME user (which would not be possible to obtain with a simple mesh, at least not when you have >20 watchers)

**Comment by akshat7101999** [comment]
Question is: WHY all these users do not see the user1 (the master user) ?
They are all seeing master only ...... **user4 can see user4 and user2**--- By this i dont mean that user4 is getting stream of user2. By this i mean that user 2 is getting stream from master and then user2 is forwarding the received stream to user4. Basically user2 is relaying the master stream to user4. So user 4 instead of connecting directly to master is getting master's stream only but from user 2. I should have cleared it in the beginning.

**Comment by yarekc** [comment]
ok, but still not unclear.

1) So the goal of this project is to have 1ToN broadcast.
Let say user1 is braodcatser, so user2, user3, user4, userN can watch him (N can be >20)
Is that correct ?

2) user2, user3, user4, userN  cannot watch user1 until... they stream themselves and thus they allow access to microphone and webcam. (if user2 does not accept webcam access, he has BLACK SCREEN)

Imagine that if you want to watch a TV stream on your PC, you have to enable access to ... your cam and microphone !
Would you accept that ?

**Comment by akshat7101999** [comment]
Yes that is correct !!

Ya i agree but this project is just a good starting point i would say. It is far from perfect. My internship is over and the rest are building on top of it. One potential solution to the issue you just raised is to remove your stream once ICE candidates are exchanged which can be incorporated in webrtc.js file so that there is only one way streaming. 
The point is that it is just a demonstration and there is quite a room for improvement.

**Comment by yarekc** [comment]
ok
I think you really should work on that project : it can be a real success ... if it works in real life conditions.

**Comment by akshat7101999** [comment]
Thanks i will try my best :)
